text/0043-reproducible-builds.md

```diff
@@ -0,0 +1,55 @@
+# Expanding the Criteria for Reproducible Builds
+
+## Summary
+
+Given the same set of inputs and without leveraging caching, Paketo buildpacks
+should reproducibly build images.
+
+## Motivation
+
+The current set of Paketo buildpacks do a reasonably good job of helping
+developers to ship reproducible build artifacts. They reuse layers where
+possible, and don't modify files in ways that aren't reproducible. So, for
+users taking advantage of the caching behaviors of CNBs, the reproducible build
+story is largely solved.
+
+However, we can do better in some other cases. Specifically, given a codebase
+that is not changing, a set of buildpacks that are not changing, and stacks
+that are not changing, we should be able to reproduce the exact same image on
+subsequent runs without leveraging the caching features of CNB.
+
+Unfortunately, there are some cases where we've made choices that prevent this
+type of reproducible build. These cases should be removed or minimized to the
+greatest extent possible.
+
+## Detailed Explanation
+
+Many of the buildpacks contain behavior that results in non-reproducible images
+due to how they treat layer metadata. Specifically, they include a `built_at`
+metadata field that includes a timestamp for the instant that this layer was
+built.
+
+Including this type of field means that two subsequent runs of a build with
+identical inputs will produce different images.
+
+## Alternatives
+
+Alternatively, we could assert that Paketo buildpacks will not be making an
+effort to enable this type of reproducible build. However, this path doesn't
+seem reasonable. Supporting these types of reproducible builds is obviously
+possible given different choices that are not fundamental to the value
+proposition of buildpacks.
+
+## Implementation
+
+We should remove the `built_at` fields from all buildpack layers that include
+them. In our test suites that leverage this field to assert layer reuse, we
+should instead prefer to compare layer SHA values across rebuilds.
```

Comment:
> In our test suites that leverage this field to assert layer reuse, we should instead prefer to compare layer SHA values across rebuilds.

Maybe I am missing something... but if the layer is truly reproducible won't the diffID be the same regardless of whether the buildpack rebuilt or reused the layer? This desirable and reproducible behavior means the the layer SHA is not useful for this specific type of assertion.

Instead we could do one of:
1. test layer reuse decisions only at the unit level
2. toggle build_at metadata with an flag that is off by default, use this for testing

I lean towards option one

Reply:
You are absolutely right that this assertion alone is not enough to tell that a layer was not rebuilt. In fact, that is really the whole purpose of the RFC, to allow a real rebuild to produce a bit-for-bit duplicate of a previous build. This is why we had relied on the `built_at` field to provide a stronger assertion of the behavior of the buildpack from the outside.

We already have test coverage of layer reuse at the unit level. Additionally, after making these proposed changes, we would still have reasonable coverage through an integration test, albeit of a weaker form where we make two different assertions (this SHA-match assertion, and an assertion on log output indicating layer reuse) that allow us to say that this layer was reused rather than rebuilt.